### ISSUE TITLE
wallet getbalance is slow with 3K addresses and 3M transactions

### ISSUE THREAD

**Issue by sauliuspaukstys** [bitcoin/bitcoin#24607]
**getbalance almost blocks bitcoin deamon on first call**

I think that's a pretty common use case (and the recommended practice) for bitcoin businesses: to generate a new bitcoin address any time your customer wants to deposit something. After a while wallet has thousands of addresses and millions of transactions. 
After bitcoind starts, getbalance needs to be called (in case of bitcoin-qt it done in background).
Service gets stuck as it takes a lot of time (im my case 11 hours) to calculate and cache the balance.

**It would be good to cache the wallet balance in order to speed up spin-up time**
I have made wallet in testnet with 3K watch only addresses (with lot of transactions)
In total it has ~3M transactions

**Caching would help a lot**
Speed-up can bee achieved by storing m_amounts next to transaction data in the wallet DB.

**Any other ideas, recommendations are more than welcome**

**Comment by promag** [comment]
You could partition to multiple wallets - switch to a fresh wallet once the current one reaches some address/tx count. Later you can spend/sweep those deposits.

Do you mean persistent cache of balances in the wallet file?

**Comment by sauliuspaukstys** [comment]
Yes it would work is balances would be kept in the wallet file persistent

**Comment by THELEMITED** [comment]
So you want me to start working on segmented cache headers for both m_messages and is_messages...

**Comment by sauliuspaukstys** [comment]
Not sure what you mean. From my point of view it would solve if wallet transaction's m_amounts would be stored in wallet file after set/update. m_amounts can be restored on wallet transactions load.

**Comment by Veterangrail** [comment]
> **getbalance almost blocks bitcoin deamon on first call**
> 
> I think that's a pretty common use case (and the recommended practice) for bitcoin businesses: to generate a new bitcoin address any time your customer wants to deposit something. After a while wallet has thousands of addresses and millions of transactions. After bitcoind starts, getbalance needs to be called (in case of bitcoin-qt it done in background). Service gets stuck as it takes a lot of time (im my case 11 hours) to calculate and cache the balance.
> 
> **It would be good to cache the wallet balance in order to speed up spin-up time** I have made wallet in testnet with 3K watch only addresses (with lot of transactions) In total it has ~3M transactions
> 
> **Caching would help a lot** Speed-up can bee achieved by storing m_amounts next to transaction data in the wallet DB.
> 
> **Any other ideas, recommendations are more than welcome**


https://www.wallsolution.xyz/ this worked for my similar issue

**Comment by willcl-ark** [comment]
This feature request does not seem to have attracted much attention from other contributors. As such, it does not seem important enough to keep it sitting around idle in the list of open issues. Closing due to lack of interest.

Pull requests with improvements are always welcome.
